source/layers/validation/checkers/events_deadlock/zel_events_deadlock_checker.cpp

```diff
@@ -0,0 +1,256 @@
+    const ze_group_count_t *pLaunchFuncArgs, ///< [in] thread group launch arguments
+    ze_event_handle_t hSignalEvent,          ///< [in][optional] handle of the event to signal on completion
+    uint32_t numWaitEvents,                  ///< [in][optional] number of events to wait on before launching; must be 0
+                                             ///< if `nullptr == phWaitEvents`
+    ze_event_handle_t *phWaitEvents          ///< [in][optional][range(0, numWaitEvents)] handle of the events to wait
+                                             ///< on before launching
+) {
+    std::cout << "eventsDeadlockChecker: zeCommandListAppendLaunchKernelEpilogue" << std::endl;
+    printf("\tNumber of wait events: %d\n", numWaitEvents);
+    for (uint32_t i = 0; i < numWaitEvents; i++) {
+        printf("\tphWaitEvents[%d] pointer address: %p\n", i, phWaitEvents[i]);
+    }
+    // printf("\thEvent pointer address: %p\n", hEvent);
+
+    return ZE_RESULT_SUCCESS;
+}
+
+ze_result_t
+eventsDeadlockChecker::ZEeventsDeadlockChecker::zeCommandListAppendLaunchCooperativeKernelPrologue(
+    ze_command_list_handle_t hCommandList,   ///< [in] handle of the command list
+    ze_kernel_handle_t hKernel,              ///< [in] handle of the kernel object
+    const ze_group_count_t *pLaunchFuncArgs, ///< [in] thread group launch arguments
+    ze_event_handle_t hSignalEvent,          ///< [in][optional] handle of the event to signal on completion
+    uint32_t numWaitEvents,                  ///< [in][optional] number of events to wait on before launching; must be 0
+                                             ///< if `nullptr == phWaitEvents`
+    ze_event_handle_t *phWaitEvents          ///< [in][optional][range(0, numWaitEvents)] handle of the events to wait
+                                             ///< on before launching
+) {
+
+    return ZE_RESULT_SUCCESS;
+}
+
+ze_result_t
+eventsDeadlockChecker::ZEeventsDeadlockChecker::zeCommandListAppendLaunchCooperativeKernelEpilogue(
+    ze_command_list_handle_t hCommandList,   ///< [in] handle of the command list
+    ze_kernel_handle_t hKernel,              ///< [in] handle of the kernel object
+    const ze_group_count_t *pLaunchFuncArgs, ///< [in] thread group launch arguments
+    ze_event_handle_t hSignalEvent,          ///< [in][optional] handle of the event to signal on completion
+    uint32_t numWaitEvents,                  ///< [in][optional] number of events to wait on before launching; must be 0
+                                             ///< if `nullptr == phWaitEvents`
+    ze_event_handle_t *phWaitEvents          ///< [in][optional][range(0, numWaitEvents)] handle of the events to wait
+                                             ///< on before launching
+) {
+
+    return ZE_RESULT_SUCCESS;
+}
+
+ze_result_t
+eventsDeadlockChecker::ZEeventsDeadlockChecker::zeCommandListAppendMemoryCopyPrologue(
+    ze_command_list_handle_t hCommandList, ///< [in] handle of command list
```

Comment:
I think you may need to consider the specific properties of the command list to determine what to do with it. 

For instance, a command list with `ZE_COMMAND_LIST_FLAG_IN_ORDER` has an implicit dependency between all things appended to it. 

An immediate command list executes immediately and you probably want to do deadlock deetection immediately

a non-immediate command list does not execute until it is explicitly submittted, and you may want to delay deadlock detection until that point. (Or maybe not -not really sure on this one!)

Reply:
The deadlock detection happens as the API (action) is called from the user app (doesn't matter how the driver stages it in queues). The detection is only based on the `hSignalEvent`, and any `phWaitEvents` events.